hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -2118,6 +2188,32 @@ def _pop_ir_tree_node(self, ir_type: IRTypeName, kwargs: IRKWargsType) -> IRNode
         if not ir_value_permitted(node.value, node.ir_type, kwargs):
             self.mark_invalid()
 
+        if forced is not None:
+            # if we expected a forced node but are instead returning a non-forced
+            # node, something has gone terribly wrong. If we allowed this combination,
+            # we risk violating core invariants that rely on forced draws being,
+            # well, forced to a particular value.
+            #
+            # In particular, this can manifest while shrinking. Consider the tree
+            #   [boolean True [forced] {"p": 0.5}]
+            #   [boolean False {"p": 0.5}]
+            #
+            # and the shrinker tries to reorder these to
+            #   [boolean False {"p": 0.5}]
+            #   [boolean True [forced] {"p": 0.5}].
+            #
+            # However, maybe we got lucky and the non-forced node is returning
+            # the same value that was expected from the forced draw. We lucked
+            # into an aligned tree in this case and can let it slide.
+            if not node.was_forced and not ir_value_equal(ir_type, forced, node.value):
+                self.mark_invalid()
+
+            # similarly, if we expected a forced node with a certain value, and
+            # are returning a forced node with a different value, this is an
+            # equally bad misalignment.
+            if node.was_forced and not ir_value_equal(ir_type, forced, node.value):
+                self.mark_invalid()
```

Comment:
you know, I was just writing up a review comment about how I'm baffled that this was never a problem for buffers, and then I realized that `draw_bits(forced=v)` just returns v and ignores the buffer. We could conceivably do the same thing here and ignore the fact that we have an ir tree that wants to provide a different value for the forced node?

This is again the tension of bailing out early for misaligned trees, or letting it play out. This time I have absolutely no intuition which is better in the long run for `forced`!

Reply:
Our options are pretty clearly (1) return the forced value regardless of the IR tree, or (2) `.mark_invalid()` and halt.

Since we might have spent a fair bit of time running user code (composed with strategies or in the test proper), I expect that the former is substantially more efficient.  Plausibly there's something clever the shrinker could do based on this information, but I think that's a concern for well after we finish the IR migration.